src/util/Keycloak.ts

```diff
@@ -7,4 +9,25 @@ const keycloak = Keycloak({
     clientId: process.env.REACT_APP_KEYCLOAK_CLIENTID!,
 });
 
+/**
+ * Refresh token minimum validity in seconds.
+ *
+ * Should be less than Access Token Lifespan in Keycloak.
+ */
+const KEYCLOAK_REFRESH_VALIDITY_TIME = 100;
+
+keycloak.onTokenExpired = () => {
+    keycloak.updateToken(KEYCLOAK_REFRESH_VALIDITY_TIME).then((refreshed) => {
```

Comment:
According to the docs, the param minValidity of updateToken function is:
> If the token expires within minValidity seconds (minValidity is optional, if not specified 5 is used) the token is refreshed.

If you check out the Keycloak code, the onTokenExpired is triggered only after the token has actually expired (which is unfortunate), so leaving out the validity time would work as well: https://github.com/keycloak/keycloak/blob/66dfa32cd569a7416de21b4dc04db212e8fccce5/adapters/oidc/js/src/main/resources/keycloak.js#L1017

One consequence of Keycloak behaving this way is that there may be a situation where an access token is already expired, but the new one is not fetched yet. If there is a running XHR API request at that time, it will 401. Standard approach is to trigger the expiration notification e.g. one minute prior to the actual expiration so that there is enough time to refresh the token, but strangely enough, at least as far as I can tell from the Keycloak JS source code, the notification is triggered only after the token is expired already.

Reply:
Good point. However, the TS definition of `updateToken` requires the `minValidity` parameter value, so I had to provide a value there.

Reply:
Yeah, the library is not of the best quality... So I guess the implementation is good enough, considering the circumstances.